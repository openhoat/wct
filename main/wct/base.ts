import { html } from 'common-tags'

export { html }

export type WebComponentState = Record<string, unknown>

export type WebComponentTemplate = <T extends WebComponent = WebComponent>(
  this: T,
) => string

export type WebComponentSpec<
  T extends WebComponentState | undefined = undefined,
> = {
  css?: string
  initialState?: T
  tagName: string
  template?: WebComponentTemplate
}

export const DefaultWebComponentTemplate = function <
  T extends WebComponent = WebComponent,
>(this: T) {
  return html`<pre title="Not yet implemented!">${webComponentLogTag(this.componentClassName)}</pre>`
}

export type WebComponentRegisterOptions = {
  log?: boolean
}

export const webComponentLogTag = (webComponentClassName: string) =>
  `[${webComponentClassName}]`

export interface WebComponent<T extends WebComponentState = WebComponentState>
  extends HTMLElement {
  readonly componentClassName: string
  readonly state: Partial<T> | undefined
  getState(name: string): unknown
  render(): void
  setState(name: string, value: unknown): void
}

export class BaseWebComponent<T extends WebComponentState = WebComponentState>
  extends HTMLElement
  implements WebComponent<T>
{
  _state?: Partial<T>
  css?: string
  componentClassName: string = BaseWebComponent.name
  log = false

  get logTag() {
    return webComponentLogTag(this.componentClassName)
  }

  get state(): Partial<T> | undefined {
    return this._state
  }

  constructor() {
    super()
    if (this.log) {
      console.log(`Building ${this.logTag}…`)
    }
    const template = document.createElement('template')
    this.attachShadow({ mode: 'open' }).appendChild(
      template.content.cloneNode(true),
    )
    if (this.log) {
      console.log(`${this.logTag} built.`)
    }
  }

  template?(context: { state?: Partial<T> }): string

  getState(name: string) {
    return this._state ? this._state[name] : undefined
  }

  removeState(name: string, options?: { render?: boolean }) {
    if (this.state === undefined) {
      return false
    }
    if (this.getState(name) === undefined) {
      return false
    }
    const { [name]: __, ...newState } = this.state
    this._state = newState as Partial<T>
    const { render = true } = options || {}
    if (render) {
      this.render()
    }
    return true
  }
  setState(name: string, value: unknown, options?: { render?: boolean }) {
    if (this.getState(name) === value) {
      return false
    }
    const newState = { ...this.state, [name]: value } as Partial<T>
    if (this.log) {
      console.log(`New state for ${this.logTag}.`)
    }
    this._state = newState
    const { render = true } = options || {}
    if (render) {
      this.render()
    }
    return true
  }

  render(): void {
    if (this.log) {
      console.log(`Rendering ${this.logTag}…`)
    }
    if (!this.shadowRoot) {
      return
    }
    if (this.log) {
      console.log(`Building ${this.logTag} template…`)
    }
    const context = { state: this._state }
    const css = this.css
    const htmlContent = this.template
      ? css
        ? html`
              <style>${css}</style>
              ${this.template(context)}
            `
        : this.template(context)
      : ''
    if (this.log) {
      console.log(`${this.logTag} template: "${htmlContent}".`)
    }
    this.shadowRoot.innerHTML = htmlContent
    if (this.log) {
      console.log(`${this.logTag} rendered.`)
    }
  }
}
