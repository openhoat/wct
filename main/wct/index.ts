import { toPascalCase } from '@/utils/helper'
import {
  BaseWebComponent,
  DefaultWebComponentTemplate,
  type WebComponent,
  type WebComponentRegisterOptions,
  type WebComponentSpec,
  webComponentLogTag,
} from './base'

export const defineWebComponent = <T extends WebComponent = WebComponent>(
  spec: WebComponentSpec<T['state']> & Pick<WebComponentRegisterOptions, 'log'>,
): void => {
  const {
    css,
    initialState,
    log,
    tagName,
    template = DefaultWebComponentTemplate,
  } = spec
  const webComponentClassName = toPascalCase(tagName)
  if (log) {
    console.log(`Registering ${webComponentLogTag(webComponentClassName)}â€¦`)
  }
  const { [webComponentClassName]: webComponentClass } = {
    [webComponentClassName]: class extends BaseWebComponent {
      constructor() {
        super()
        if (log !== undefined) {
          this.log = log
        }
        this.css = css
        this.componentClassName = webComponentClassName
        this.template = template
        if (initialState !== undefined) {
          for (const [name, value] of Object.entries(initialState)) {
            this.setState(name, value, { render: false })
          }
        }
        this.render()
      }
    },
  }
  customElements.define(tagName, webComponentClass)
  if (log) {
    console.log(`${webComponentLogTag(webComponentClassName)} registered.`)
  }
}

export const getWebComponent = <T extends WebComponent = WebComponent>(
  selector: string,
  root?: string,
): T | undefined =>
  (root !== undefined
    ? document.querySelector(root)?.shadowRoot?.querySelector<T>(selector)
    : document.querySelector<T>(selector)) ?? undefined

export const registerComponentSpec = <T extends WebComponent = WebComponent>(
  spec: WebComponentSpec<T['state']>,
  options?: WebComponentRegisterOptions,
) => {
  return defineWebComponent<T>({
    ...options,
    ...spec,
  })
}

export const registerComponentSpecs = <T extends WebComponent = WebComponent>(
  specs: WebComponentSpec<T['state']>[],
  options?: WebComponentRegisterOptions,
) => {
  for (const spec of specs) {
    registerComponentSpec(spec, options)
  }
}
