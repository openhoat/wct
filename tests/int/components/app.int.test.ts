import { getWebComponent, registerComponentSpec } from '@/wct'
import type { WebComponentSpec } from '@/wct/base'
import { html } from 'common-tags'

describe('App Component', () => {
  const spec: WebComponentSpec = {
    tagName: 'h-app',
    template: () => html`
          <h-header></h-header>
          <h-todo-list></h-todo-list>
          <h-footer id="footer"></h-footer>
        `,
  }
  beforeAll(() => {
    document.body.innerHTML = html`<h-app></h-app>`
    registerComponentSpec(spec)
  })
  it('should render app component', () => {
    // Given
    const component = getWebComponent('h-app')
    // When
    const result = component?.shadowRoot?.innerHTML
    // Then
    expect(result).toMatchSnapshot()
  })
})
