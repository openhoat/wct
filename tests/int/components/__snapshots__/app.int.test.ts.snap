// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render app component 1`] = `
"<h-header></h-header>
<h-todo-list></h-todo-list>
<h-footer id="footer"></h-footer>"
`;
